js_modules/dagster-ui/packages/ui-core/src/overview/assets/AssetTimelineDataFetcher.tsx

```diff
@@ -0,0 +1,414 @@
+              const runs = groupedRuns[groupKey]?.runs ?? [];
+              runsCache.current.groups[groupKey] = {
+                runs: Array.from(runs),
+                loading: false,
+              };
+              assetKeys.forEach((key) => {
+                const stringAssetKey = tokenForAssetKey(key);
+                runsCache.current.assets[stringAssetKey] = {
+                  runs: groupedRuns[groupKey]?.assets[stringAssetKey] ?? [],
+                  loading: false,
+                };
+              });
+            });
+            triggerListeners();
+          });
+        }),
+      );
+      isInProgress = false;
+    }
+
+    const interval = setInterval(tick, POLL_INTERVAL);
+    setTimeout(() => tick(true), 100);
+    return () => {
+      isCanceled = true;
+      clearInterval(interval);
+    };
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [client, didCodeLocationNodesChange, didRangeMeaningfullyChange]);
+
+  const value = React.useMemo(() => {
+    return {
+      subscribeToAssetData: (key: AssetKeyInput, listener: Listener) => {
+        const stringKey = tokenForAssetKey(key);
+        assetSubscriptions.current[stringKey] = listener;
+        const data = runsCache.current.assets[stringKey];
+        if (data) {
+          listener(data);
+        }
+        return () => {
+          delete assetSubscriptions.current[stringKey];
+        };
+      },
+      subscribeToGroupData: (groupName: string, listener: Listener) => {
+        groupSubscriptions.current[groupName] = listener;
+        const data = runsCache.current.groups[groupName];
+        if (data) {
+          listener(data);
+        }
+        return () => {
+          delete groupSubscriptions.current[groupName];
+        };
+      },
+      subscribeToCodeLocationData: (codeLocation: string, listener: Listener) => {
+        codeLocationSubscriptions.current[codeLocation] = listener;
+        const data = runsCache.current.locations[codeLocation];
+        if (data) {
+          listener(data);
+        }
+        return () => {
+          delete codeLocationSubscriptions.current[codeLocation];
+        };
+      },
+    };
+  }, []);
+  return <DataFetcherContext.Provider value={value}>{children}</DataFetcherContext.Provider>;
+};
+
+function groupRunsByGroup(runs: RunWithAssetsFragment[], assetToGroup: Record<string, string>) {
+  const groupedRuns: Record<
+    string,
+    {
+      assets: Record<string, RunWithAssetsFragment[]>;
+      runs: Set<RunWithAssetsFragment>;
+    }
+  > = {};
+
+  runs.forEach((run) => {
+    const {assets} = run;
+    assets.forEach((asset) => {
+      const stringKey = tokenForAssetKey(asset.key);
+      const group = assetToGroup[stringKey]!;
+      groupedRuns[group] = groupedRuns[group] || {
+        assets: {},
+        runs: new Set(),
+      };
+      groupedRuns[group]!.runs.add(run);
+      groupedRuns[group]!.assets[stringKey] = groupedRuns[group]!.assets[stringKey] || [];
```

Comment:
I'd vote to make both of these new Set() or both of them new arrays maybe? Depends on the usage i guess